File: java/src/test/java/com/unascribed/flexver/FlexVerBasicTest.java
package com.unascribed.flexver;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import com.unascribed.flexver.FlexVerComparator.NumericVersionComponent;
import com.unascribed.flexver.FlexVerComparator.SemVerPrereleaseVersionComponent;
import com.unascribed.flexver.FlexVerComparator.VersionComponent;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class FlexVerBasicTest {
	public static final String[] ENABLED_TESTS = { "test_vectors.txt", "large.txt" };

	@ParameterizedTest(name = "{0} {2} {1}")
	@MethodSource("getEqualityTests")
	public void testEquality(String a, String b, Ordering expected) {
		System.out.println(formatDecomposedString(a)+" "+expected+" "+formatDecomposedString(b));
		
		Ordering c = Ordering.fromComparison(FlexVerComparator.compare(a, b));
		Ordering c2 = Ordering.fromComparison(FlexVerComparator.compare(b, a));

		// When inverting the operands we're comparing, the ordering should be inverted too
		assertEquals(c2.invert(), c, "Comparison method violates its general contract! ("+a+" <=> "+b+" is not commutative)");

		assertEquals(expected, c, "Ordering.fromComparison produced "+a+" "+c+" "+b);
	}

	private String formatDecomposedString(String str) {
		StringBuilder sb = new StringBuilder("[");
		for (VersionComponent c : FlexVerComparator.decompose(str)) {
			int color = 90;
			if (c instanceof NumericVersionComponent) {
				color = 96;
			} else if (c instanceof SemVerPrereleaseVersionComponent) {
				color = 93;
			} else {
				color = 95;
			}
			sb.append("\u001B[").append(color).append("m");
			sb.append(c.toString());
		}
		if (str.contains("+")) {
			sb.append("\u001B[90m");
			sb.append(str.substring(str.indexOf('+')));
		}
		sb.append("\u001B[0m]");
		return sb.toString();
	}

	public static Stream<Arguments> getEqualityTests() throws IOException {
		Path testsFolder = new File("../test").toPath().toAbsolutePath();
		List<String> lines = new ArrayList<>();

		for (String test : ENABLED_TESTS) {
			lines.addAll(Files.readAllLines(testsFolder.resolve(test)));
		}

		return lines.stream()
			.filter(line -> !line.startsWith("#"))
			.filter(line -> !line.isEmpty())
			.map(line -> {
				String[] split = line.split(" ", -1);
				if (split.length != 3) throw new IllegalArgumentException("Line formatted incorrectly, expected 2 spaces: "+line);
				return Arguments.of(split[0], split[2], Ordering.fromString(split[1]));
			});
	}
}

File: csharp/FlexVerTests/FlexVerBasicTest.cs
using FlexVer;

namespace FlexVerTests;

public class Tests
{
	private const string TestDir = "";
	private static readonly string[] EnabledTests = { "test_vectors.txt", "large.txt" };

	[Test]
	[TestCaseSource(nameof(GetEqualityTests))]
	public void TestEquality_StaticCompare(string a, string b, Ordering expectedOrdering)
	{
		Ordering c = OrderingExtensions.FromComparison(FlexVerComparer.Compare(a, b));
		Ordering c2 = OrderingExtensions.FromComparison(FlexVerComparer.Compare(b, a));

		Assert.That(c, Is.EqualTo(c2.Invert()), $"Comparison method violates its general contract! ({a} <=> {b} is not commutative)");
		Assert.That(c, Is.EqualTo(expectedOrdering), $"OrderingExtensions.FromComparison produced {a} {c} {b}");
	}

	[Test]
	[TestCaseSource(nameof(GetEqualityTests))]
	public void TestEquality_DefaultComparerInstance(string a, string b, Ordering expectedOrdering)
	{
		Ordering c = OrderingExtensions.FromComparison(FlexVerComparer.Default.Compare(a, b));
		Ordering c2 = OrderingExtensions.FromComparison(FlexVerComparer.Default.Compare(b, a));

		Assert.That(c, Is.EqualTo(c2.Invert()), $"Comparison method violates its general contract! ({a} <=> {b} is not commutative)");
		Assert.That(c, Is.EqualTo(expectedOrdering), $"OrderingExtensions.FromComparison produced {a} {c} {b}");
	}

	[Test]
	[TestCase(null, null, Ordering.Equal)]
	[TestCase(null, "1.0.0", Ordering.Less)]
	[TestCase("1.0.0", null, Ordering.Greater)]
	[TestCase("1.0.0", "1.0.0", Ordering.Equal)]
	public void TestEquality_DefaultComparerInstance_HandlesClrNulls(string? a, string? b, Ordering expectedOrdering)
	{
		Ordering c = OrderingExtensions.FromComparison(FlexVerComparer.Default.Compare(a, b));
		Ordering c2 = OrderingExtensions.FromComparison(FlexVerComparer.Default.Compare(b, a));

		Assert.That(c, Is.EqualTo(c2.Invert()), $"Comparison method violates its general contract! ({a} <=> {b} is not commutative)");
		Assert.That(c, Is.EqualTo(expectedOrdering), $"OrderingExtensions.FromComparison produced {a} {c} {b}");
	}


	internal static IEnumerable<object?[]> GetEqualityTests()
	=> EnabledTests
		.SelectMany(testFileName => File.ReadAllLines(Path.Join(TestDir, testFileName)))
		.Where(line => !line.StartsWith('#'))
		.Where(line => !string.IsNullOrWhiteSpace(line))
		.Select(line =>
		{
			var split = line.Split(' ');
			if (split.Length != 3) throw new ArgumentException($"Line formatted incorrectly, expected 2 spaces: {line}");
			return new object?[] { split[0], split[2], OrderingExtensions.FromChar(split[1]) };
		});
}
